File: ConsoleChat/src/consolechatserver-linux/src/Framework.cpp
/*
 * NetDrone Engine
 * Copyright © 2022 Origin Studio Inc.
 *
 */

#include "Framework.h"

#include "CommonHeader.h"

template<> CFramework* CSingleton< CFramework >::ms_Instance = NULL;

bool
CFramework::Initialize(const char* szConfigPath_, const char* szPackageName_, INT iServerId_) {
	g_kConnectorProvider.Register(NETWORK_TCP_CLIENT, CTcpConnector::New);
	g_kConnectorProvider.Register(NETWORK_UDP_CLIENT, CUdpConnector::New);
	g_kConnectorProvider.Register(NETWORK_TCP_SERVER, CTcpListener::New);
	g_kConnectorProvider.Register(NETWORK_UDP_SERVER, CUdpListener::New);

	g_kRoomProvider.Register(CRoomEx::New);
	g_kPlayerProvider.Register(CPlayerEx::New);

	m_bInitialized = false;

	if(Action("initializing configure", g_kCfgMgr.Initialize(szConfigPath_, szPackageName_, iServerId_))) {
		if(Action("initializing protocol", g_kNetMgr.CheckProtocol())) {
			INT iMaxUser = g_kCfgMgr.GetMaxConnection();

			//TRACE("max user: " << iMaxUser);

			if(getdtablesize() < iMaxUser) {
				OUTPUT("["<<g_kTick.GetTime()<<"] socket is not enough: "<<getdtablesize()<<" : "<<iMaxUser);
				return false;
			}

			if(Action("initializing server", g_kServerMgr.Initialize())) {
				if(Action("initializing channel", g_kChannelMgr.Initialize())) {
					if(10 <= iMaxUser) {
						g_kChannelMgr.GetServerInfo().SetId(g_kCfgMgr.GetServerId());
						g_kChannelMgr.GetServerInfo().SetMaxUser(iMaxUser);

						// 채널별 최대 동시 접속자수를 설정합니다.
						for(UINT i = 0; i < (UINT)(iMAX_CHANNEL_LIST); ++i) {
							UINT uiChannelId = i+1;
							UINT uiChannelMaxUser = iMaxUser / iMAX_CHANNEL_LIST;

							g_kChannelMgr.GetChannelInfo(i).SetId(uiChannelId);
							g_kChannelMgr.GetChannelInfo(i).SetMaxUser(uiChannelMaxUser);
						}

						g_kChannelMgr.SetMaxChannel(iMAX_CHANNEL_LIST);

						if(Action("initializing network", g_kNetMgr.Initialize(iMaxUser))) {
							if(Action("initializing player's trunk created", g_kUnitMgr.NewTrunk(UNIT_PLAYER, iMaxUser+10))) {
								m_bInitialized = true;
								m_bDoing = true;

								char szFileLogPath[iFILE_PATH_LEN+1] = { 0, };
								snprintf(szFileLogPath, iFILE_PATH_LEN, "%s/%d", g_kCfgMgr.GetLogPath(), 1000000 + g_kCfgMgr.GetServerId());

								char szHackLogPath[iFILE_PATH_LEN+1] = { 0, };
								snprintf(szHackLogPath, iFILE_PATH_LEN, "%s/%d", g_kCfgMgr.GetLogPath(), 2000000 + g_kCfgMgr.GetServerId());

								g_kFileLog = new CLogFile(szFileLogPath);
								g_kHackLog = new CLogFile(szHackLogPath);

								Action("Starting service", Startup());
							}
						}
					} else {
						OUTPUT("["<<g_kTick.GetTime()<<"] max user < 10");
					}
				}
			}
		}
	}
	return m_bInitialized;
}

bool
CFramework::Release() {
	if(m_bInitialized) {
		Action("shutdown immediately", Shutdown());

		m_bInitialized = false;
		m_bDoing = false;

		//OUTPUT("["<<g_kTick.GetTime()<<"] ["<<getpid()<<"] server shutdown now");

		CAllocator< CNativeCommand >::FreeAllPool();
		CAllocator< CExtendCommand >::FreeAllPool();
		CAllocator< CSendCommand >::FreeAllPool();
		CAllocator< CTcpPacketBuffer >::FreeAllPool();
		CAllocator< CUdpPacketBuffer >::FreeAllPool();

		g_kUnitMgr.ReleaseInstance();
		g_kNetMgr.ReleaseInstance();
		g_kChannelMgr.ReleaseInstance();
		g_kServerMgr.ReleaseInstance();

		g_kCfgMgr.ReleaseInstance();

		g_kRoomProvider.ReleaseInstance();
		g_kPlayerProvider.ReleaseInstance();
		g_kConnectorProvider.ReleaseInstance();
	}
	return true;
}

bool
CFramework::Update() {
	g_kTick.Update();
	g_kNetMgr.Update();
	g_kServerMgr.Update();

	return true;
}

bool
CFramework::Startup() {
	if(m_bInitialized) {
		if(g_kNetMgr.StartThread()) {
			//TRACE("network manager: sender thread: start");
		} else {
			//TRACE("network manager: sender thread: not used");
		}

		FILELOG(LE_SERVER_START);
		return true;
	}
	return false;
}

bool
CFramework::Shutdown() {
	if(m_bInitialized) {
		TRACE("[" << g_kTick.GetTime() << "] [" << g_kCfgMgr.GetServerId() << "] shutdown doing.....");

		g_kChannelMgr.Shutdown();
		time_t tmTimeout = g_kTick.GetTime() + 300;
		while(0 < g_kChannelMgr.GetServerInfo().GetUser()) {
			Update();
			TRACE("timeout: " << tmTimeout << ", now: " << g_kTick.GetTime() << ", user: " << g_kChannelMgr.GetServerInfo().GetUser());

			usleep(0);
			if(tmTimeout < g_kTick.GetTime()) {
				break;
			}
		}

		g_kServerMgr.Shutdown();
		tmTimeout = g_kTick.GetTime() + 3;
		while(g_kServerMgr.IsAvailable()) {
			Update();
			TRACE("timeout: " << tmTimeout << ", now: " << g_kTick.GetTime());

			usleep(0);
			if(tmTimeout < g_kTick.GetTime()) {
				break;
			}
		}

		usleep(1000);
		g_kNetMgr.Update();
		usleep(1000);

		TRACE("[" << g_kTick.GetTime() << "] [" << g_kCfgMgr.GetServerId() << "] shutdown doing: all user were disconnected: ok: ");

		if(g_kNetMgr.CancelThread()) {
			//TRACE("network manager: sender thread: cancel");
		} else {
			//TRACE("network manager: sender thread: not used");
		}

		FILELOG(LE_SERVER_STOP);
		usleep(1000);
		return true;
	}
	return false;
}

/* EOF */

File: ConsoleChat/src/consolechatclient/Framework.cs
/*
 * NetDrone Engine
 * Copyright © 2022 Origin Studio Inc.
 *
 */

using System;
using System.Text;
using System.Net.Sockets;

namespace CompatibilityStandards {
	#region User-Defined Types
	using UINT = System.UInt32;
	using BYTE = System.Byte;
	using SBYTE = System.SByte;
	using WORD = System.UInt16;
	using DWORD = System.UInt32;
	using QWORD = System.UInt64;
	using ULONG = System.UInt32;
	using ULONG32 = System.UInt32;
	using ULONG64 = System.UInt64;
	using CHAR = System.Byte;
	using INT = System.Int32;
	using INT16 = System.Int16;
	using INT32 = System.Int32;
	using INT64 = System.Int64;
	using UINT16 = System.UInt16;
	using UINT32 = System.UInt32;
	using UINT64 = System.UInt64;
	using LONG32 = System.Int32;
	using LONG64 = System.Int64;
	using FLOAT = System.Single;
	using DOUBLE = System.Double;
	using tick_t = System.UInt64;
	using time_t = System.UInt64;
	using size_t = System.UInt64;
	using wchar_t = System.Char;
	#endregion

	public partial class GameFramework {
		public GameFramework() {}
		~GameFramework() {}

		public static CFramework	g_kFramework		{ get { return CFramework.GetInstance(); } }

		public class CFramework : CSingleton< CFramework > {
			public CFramework() {}
			~CFramework() { Release(); }

			public bool
			Initialize() {
				if(false == m_bInitialized) {
					g_kConnectorProvider.Register((INT)NETWORK_TYPE.NETWORK_TCP_CLIENT, new CreateCallback(CTcpConnector.New));
					g_kConnectorProvider.Register((INT)NETWORK_TYPE.NETWORK_TCP_SERVER, new CreateCallback(CTcpListener.New));
					g_kConnectorProvider.Register((INT)NETWORK_TYPE.NETWORK_UDP_CLIENT, new CreateCallback(CUdpConnector.New));
					g_kConnectorProvider.Register((INT)NETWORK_TYPE.NETWORK_UDP_SERVER, new CreateCallback(CUdpListener.New));

					g_kCfgMgr.Initialize();
					g_kNetMgr.Initialize();
					g_kNetMgr.CheckProtocol();

					m_bInitialized = true;
					m_bDoing = true;

					Startup();

					return true;
				}
				return false;
			}

			public bool
			Release() {
				if(m_bInitialized) {
					Shutdown();

					m_bInitialized = false;
					m_bDoing = false;

					CAllocator< CNativeCommand >.FreeAllPool();
					CAllocator< CTcpPacketBuffer >.FreeAllPool();
					CAllocator< CUdpPacketBuffer >.FreeAllPool();

					g_kConnectorProvider.ReleaseInstance();

					g_kNetMgr.Release();
					g_kNetMgr.ReleaseInstance();
					g_kCfgMgr.ReleaseInstance();
				}
				return true;
			}

			public virtual bool
			CreateFrameworkThread(UINT uiSleepTime_ =0) {
#if _THREAD
				m_kFrameworkThread = new CFrameworkThread(this, (INT)uiSleepTime_);
				return true;
#else
				return false;
#endif
			}

			public virtual bool
			StartThread() {
#if _THREAD
				if(isptr(m_kFrameworkThread)) {
					Action("starting framework thread", m_kFrameworkThread.Start());
				}
				return true;
#else
				return false;
#endif
			}

			public virtual bool
			CancelThread() {
#if _THREAD
				if(isptr(m_kFrameworkThread)) {
					Action("stopping framework thread", m_kFrameworkThread.Cancel());
				}
				return true;
#else
				return false;
#endif
			}

			public bool
			Update() {
				if(false == m_bDoing) {
					return false;
				}

				g_kTick.Update();
				g_kNetMgr.Update();

				return true;
			}

			public bool
			Startup() {
				if(m_bInitialized) {
					if(g_kNetMgr.StartThread()) {
						//TRACE("network manager: sender thread: start");
					} else {
						//TRACE("network manager: sender thread: not used");
					}
					return true;
				}
				return false;
			}

			public bool
			Shutdown() {
				if(m_bInitialized) {
					if(g_kNetMgr.CancelThread()) {
						//TRACE("network manager: sender thread: cancel");
					} else {
						//TRACE("network manager: sender thread: not used");
					}

					Thread.Sleep(1);
					return true;
				}
				return false;
			}

			public bool		IsDoing()			{ return m_bDoing; }
			public void		SetDoing(bool o)	{ m_bDoing = o; }

			private bool	m_bInitialized = false;
			private bool	m_bDoing = true;

#if _THREAD
			private CFrameworkThread		m_kFrameworkThread = null;
#endif
		}
	}
}

/* EOF */
